### PR TITLE
Protobuf serialization

### PR DESCRIPTION
**Pull request by palourde**
## What is this change?

This PR introduces Protobuf serialization for most of our resources.

## Why is this change necessary?

Closes https://github.com/sensu/sensu-go/issues/2957

## Does your change need a Changelog entry?

Done

## Do you need clarification on anything?

Nope

## Were there any complications while making this change?

Some resources were not generated using protobuf and others stored using `types.Wrapper` will have to wait!

## Have you reviewed and updated the documentation for this change? Is new documentation required?

Nope

## How did you verify this change?

I added a specific regression test in `TestList`. I also tested deployed directly this new version with sensu-staging, and tested an upgrade scenario.

### CODE REVIEW

File: backend/store/etcd/generic_object.proto [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+option go_package = "etcd";
+option (gogoproto.populate_all) = true;
+option (gogoproto.equal_all) = true;
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+option (gogoproto.sizer_all) = true;
+option (gogoproto.testgen_all) = true;
+
+// Asset defines an asset agents install as a dependency for a check.
+message GenericObject {
```
Comment (palourde) [inline backend/store/etcd/generic_object.proto]: FYI we use this resource to test the generic store so obviously, to have it working with protobuf I had to create this file.

File: backend/store/etcd/store.go [hunk 2]
```diff
@@ -37,7 +39,7 @@ func NewStore(client *clientv3.Client, name string) *Store {
 
 // Create the given key with the serialized object.
 func Create(ctx context.Context, client *clientv3.Client, key, namespace string, object interface{}) error {
-	bytes, err := json.Marshal(object)
+	bytes, err := proto.Marshal(object.(proto.Message))
```
Comment (echlebek) [inline backend/store/etcd/store.go]: Hmm this seems a bit unsafe. Could we change the function signature?

```
func Create(ctx context.Context, client *clientv3.Client, key, namespace string, object proto.Message) error
```

File: backend/store/etcd/store_test.go [hunk 3]
```diff
@@ -213,45 +191,54 @@ func TestList(t *testing.T) {
...
 		ctx := context.WithValue(context.Background(), types.NamespaceKey, "default")
 		require.NoError(t, Create(ctx, s.client, "/sensu.io/generic/default/obj1", "default", obj1))
 
-		obj2 := &genericObject{ObjectMeta: corev2.ObjectMeta{Name: "obj2", Namespace: "acme"}}
+		// Let's encode a resources with JSON (instead of Protobuf) to ensure the
+		// List function supports both encoding format
+		obj2 := &GenericObject{ObjectMeta: corev2.ObjectMeta{Name: "obj2", Namespace: "acme"}}
 		ctx = context.WithValue(context.Background(), types.NamespaceKey, "acme")
-		require.NoError(t, Create(ctx, s.client, "/sensu.io/generic/acme/obj2", "acme", obj2))
+		// require.NoError(t, Create(ctx, s.client, "/sensu.io/generic/acme/obj2", "acme", obj2))
```
Comment (ccressent) [inline backend/store/etcd/store_test.go]: Should this line be removed altogether?

Comment (palourde) [reply backend/store/etcd/store_test.go]: Haha yes, good catch

**Pull request merged**